picard/util/pipe.py

```diff
@@ -0,0 +1,232 @@
+
+        self.is_pipe_owner: bool = False
+
+        # test if pipe is listened to even if no args provided
+        if type(args) == list:
+            if len(args) == 0:
+                args.append(self.MESSAGE_TO_IGNORE)
+        else:
+            raise ValueError("args argument MUST be a list")
+
+        if self.__is_win:
+            for arg in args:
+                if not self.send_to_pipe(arg):
+                    self.is_pipe_owner = True
+                    break
+        else:
+            if self.__pipe_exists():
+                for arg in args:
+                    if not self.send_to_pipe(arg):
+                        try:
+                            os.unlink(self.path)
+                        except Exception as e:
+                            raise ValueError(f"There was a problem with removing the old pipe:\n{e}")
+
+                        self.__create_unix_pipe()
+                        break
+            else:
+                self.__create_unix_pipe()
+
+    def __pipe_exists(self) -> bool:
+        return os.path.exists(self.path)
+
+    def __generate_filename(self) -> str:
+        prefix: str = ""
+        username: str = os.getlogin()
+        if self.__is_win:
+            prefix = "\\\\.\\pipe\\"
+        else:
+            prefix = "/tmp/"
+
+        return f"{prefix}{self.__app_name}_v{self.__app_version}_{username}_pipe_file"
+
+    def __create_unix_pipe(self) -> None:
+        os.mkfifo(self.path)
+        self.is_pipe_owner = True
+
+    def __win_sender(self, message: str) -> bool:
+        pipe = win32pipe.CreateNamedPipe(
+            self.path,
+            win32pipe.PIPE_ACCESS_DUPLEX,
+            win32pipe.PIPE_TYPE_MESSAGE | win32pipe.PIPE_WAIT,
+            self.__MAX_INSTANCES,
+            self.__OUT_BUFFER_SIZE,
+            self.__IN_BUFFER_SIZE,
+            self.__DEFAULT_TIMEOUT,
+            None)
+        try:
+            win32pipe.ConnectNamedPipe(pipe, None)
+            win32file.WriteFile(pipe, str.encode(message))
+        finally:
+            win32file.CloseHandle(pipe)
+
+        return True
+
+    def __unix_sender(self, message: str) -> bool:
+        with open(self.path, 'a') as fifo:
+            fifo.write(message)
+        return True
+
+    def send_to_pipe(self, message: str, timeout_secs: float = 1.5) -> bool:
+        __pool = concurrent.futures.ThreadPoolExecutor()
+        sender = None
+        if self.__is_win:
+            sender = __pool.submit(self.__win_sender, message)
+        else:
+            sender = __pool.submit(self.__unix_sender, message)
+
+        try:
+            if sender.result(timeout=timeout_secs):
+                return True
+        except concurrent.futures._base.TimeoutError:
+            # hacky way to kill the sender
+            self.read_from_pipe()
+
+        return False
+
+    def read_from_pipe(self, timeout_secs: float = 1.5) -> str:
+        if self.__is_win:
+            return str(self.__read_from_win_pipe(timeout_secs))
```

Comment:
Is there a reason to not do the conversion to `str` in `__read_from_win_pipe()`?

Reply:
Actually none, fixed.